### PR TITLE
iox-#1129 Multiple RouDi MVP

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Add a second reviewer for complex new features or larger refactorings
1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Notes for Reviewer

This PR adds the experimental feature to run multiple RouDi in parallel. In order to achieve this, the `--domain-id` cmd line parameter must be used when RouDi is started. On the client side the new `Node` must be used and the domain ID must be specified with the `NodeBuilder`. There is also a `IOX_DOMAIN_ID` environment variable that can be used with the `NodeBuilder`.

Currently it is not possible to run multiple nodes in parallel in the same process due to limitations in `PointerRespository`

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Relates to #1129

### CODE REVIEW

**Comment by elBoberido** [comment]
From https://github.com/eclipse-iceoryx/iceoryx/pull/2200#discussion_r1506238490

Add tests for max nodes.

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp [hunk 1]
```diff
@@ -0,0 +1,558 @@
...
+
+    RouDiEnv roudi{roudi::DEFAULT_UNIQUE_ROUDI_ID};
+
+    IOX_POSIX_CALL(unsetenv)
+    ("IOX_ROUDI_ID").failureReturnValue(-1).evaluate().expect("Unsetting environment variable works!");
+    auto node_result = RouDiEnvNodeBuilder("foo").roudi_id_from_env_or_default().create();
+
+    EXPECT_FALSE(node_result.has_error());
+}
+
```
Comment (elBoberido) [inline iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:213]: From https://github.com/eclipse-iceoryx/iceoryx/pull/2200#discussion_r1506238490

Please also test to create the maximum supported amount of nodes.

Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:217]: Done with https://github.com/eclipse-iceoryx/iceoryx/pull/2212/commits/7604df26864e8a40f8f1307b77fbe1eab03541d2 on #2212

File: iceoryx_hoofs/test/moduletests/test_memory_scope_guard.cpp [hunk 2]
```diff
@@ -147,4 +147,20 @@ TEST_F(ScopeGuard_test, MoveAssignedCallsCleanupWhenOutOfScope)
...
+
+TEST_F(ScopeGuard_test, ReleaseInhibitsTheCallOfTheCleanupFunction)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "d849da9c-5733-4ab0-ab35-8cde433343ce");
+    int hasCalledCleanup = 0;
+
+    {
+        ScopeGuard sut([&] { ++hasCalledCleanup; });
+
+        ScopeGuard::release(std::move(sut));
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_hoofs/test/moduletests/test_memory_scope_guard.cpp:159]: I tried to use the same terminology like in `unique_ptr::release`. What do you think about changing the signature to return a `iox::function`?

File: iceoryx_posh/include/iceoryx_posh/internal/runtime/shared_memory_user.hpp [hunk 3]
```diff
@@ -18,46 +18,70 @@
...
+    /// @param[in] domainId to tie the shared memory to
+    /// @param[in] segmentId of the segment for the relocatable shared memory segment address space
+    /// @param[in] managementShmSize size of the shared memory management segment
+    /// @param[in] segmentManagerAddressOffset adress of the segment manager that does the final mapping of memory in
+    /// the process
+    /// @return a 'SharedMemoryUser' instance or an 'SharedMemoryUserError' on failure
+    static expected<SharedMemoryUser, SharedMemoryUserError>
+    create(const DomainId domainId,
+           const uint64_t segmentId,
+           const size_t managementShmSize,
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/include/iceoryx_posh/internal/runtime/shared_memory_user.hpp:58]: I did not look at the data type but just did a copy & paste from existing code. I can change the data type though.

File: iceoryx_posh/source/roudi/process.cpp [hunk 4]
```diff
@@ -27,12 +27,13 @@ namespace iox
 namespace roudi
 {
 Process::Process(const RuntimeName_t& name,
+                 const DomainId domainId,
                  const uint32_t pid,
                  const PosixUser& user,
                  const HeartbeatPoolIndexType heartbeatPoolIndex,
                  const uint64_t sessionId) noexcept
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/roudi/process.cpp:34]: We once had a startup race with the following scenario:
- app tries to register at RouDi and sends a `REG`
- RouDi is busy
- app sends another `REG`
- RouDi responds to the first `REG`
- app requests a publisher but receives the response for the second `REG`

The sessionId is a timestamp created with the `REG` and used for all messages throughout the session until the next `REG`.

File: iceoryx_posh/source/roudi/roudi_cmd_line_parser.cpp [hunk 5]
```diff
@@ -53,7 +54,11 @@ CmdLineParser::parse(int argc, char* argv[], const CmdLineArgumentParsingMode cm
             std::cout << "Options:" << std::endl;
             std::cout << "-h, --help                        Display help." << std::endl;
             std::cout << "-v, --version                     Display version." << std::endl;
-            std::cout << "-u, --unique-roudi-id <INT>       Set the unique RouDi id." << std::endl;
+            std::cout << "-d, --domain-id <UINT>            Set the Domain ID." << std::endl;
+            std::cout << "                                  <UINT> 0..65536" << std::endl;
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/roudi/roudi_cmd_line_parser.cpp:58]: No, you are right. I just forgot to subtract one ... the usual off by one error :sweat_smile:

File: iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp [hunk 6]
```diff
@@ -62,10 +65,37 @@ class NodeBuilder
     IOX_BUILDER_PARAMETER(bool, shares_address_space_with_roudi, false)
 
   public:
+    /// @brief Determines which domain to use to register to a RouDi instance
+    /// @param[in] value to be used as domain ID
+    NodeBuilder&& domain_id(const uint16_t value) && noexcept;
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp:70]: I changed it but I'm not sure if I like the API since one now has to use it like `domain_id(DomainId{42})` and I would not like it to be implicit convertible.

File: iceoryx_posh/roudi_env/source/roudi_env.cpp [hunk 7]
```diff
@@ -65,17 +70,28 @@ void RouDiEnv::triggerDiscoveryLoopAndWaitToFinish() noexcept
 
 void RouDiEnv::cleanupAppResources(const RuntimeName_t& name) noexcept
 {
-    m_runtimes.eraseRuntime(name);
+    if (m_runtimes.has_value())
+    {
+        m_runtimes->eraseRuntime(name);
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/roudi_env/source/roudi_env.cpp:75]: No, the `optional` is only there to support the old way to create a runtime, i.e. the static runtime which will be a thread local one when used with the `RouDiEnv` but also have to option to not use it at all when the new way is used via the node. `m_runtimes` is also not just a runtime but a `RuntimeTestInterface` which contains all the thread local runtimes.

`eraseRuntime` does essentially the same like `cleanupRuntimes` but `eraseRuntime` does it only for the one specified and `cleanupRuntimes` for all active runtimes. Both are private methods though.

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp [hunk 8]
```diff
@@ -388,4 +498,46 @@ TEST(Node_test, NodeAndEndpointsAreContinuouslyRecreated)
...
+{
+    ::testing::Test::RecordProperty("TEST_ID", "1e527815-28d1-4a99-a9a3-cc4084018cf3");
+
+    NodeName_t node_name{"hypnotoad"};
+    ServiceDescription service_description{"all", "glory", "hypnotoad"};
+
+    constexpr uint16_t domain_id_a{13};
+    constexpr uint16_t domain_id_b{42};
+
+    RouDiEnv roudi_a{DomainId{domain_id_a}};
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:511]: There is a test in `test_roudi_env.cpp` but it panics. The use case is different though since one usually does not start RouDi in the same process like the applications but in a separate process.

If one wants to have a more graceful error handling for RouDi, one needs to adapt `IceOryxRouDiMemoryManager` to not panic in the ctor.

It is out of scope for this PR though.

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp [hunk 9]
```diff
@@ -105,6 +106,115 @@ TEST(Node_test, CreatingNodeWithInvalidNameLeadsToError)
...
+
+    RouDiEnv roudi{DomainId{42}};
+
+    constexpr int32_t OVERWRITE_ENV_VARIABLE{1};
+    IOX_POSIX_CALL(setenv)
+    ("IOX_DOMAIN_ID", "42", OVERWRITE_ENV_VARIABLE)
+        .failureReturnValue(-1)
+        .evaluate()
+        .expect("Setting environment variable works!");
+    auto node_result = RouDiEnvNodeBuilder("foo").domain_id_from_env_or(13).create();
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:169]: It is implicit verified by the node registering at RouDi and the only RouDi which is running is one with the id `42`.

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp [hunk 10]
```diff
@@ -105,6 +106,115 @@ TEST(Node_test, CreatingNodeWithInvalidNameLeadsToError)
...
+
+TEST(Node_test, CreatingNodeWithDomainIdFromEnvOrAlternativeValueWorksIfDomainIdIsNotSet)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "b071843a-a821-43b4-ac1a-e76ccafd35e0");
+
+    RouDiEnv roudi{DomainId{13}};
+
+    IOX_POSIX_CALL(unsetenv)
+    ("IOX_DOMAIN_ID").failureReturnValue(-1).evaluate().expect("Unsetting environment variable works!");
+    auto node_result = RouDiEnvNodeBuilder("foo").domain_id_from_env_or(13).create();
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:182]: Same here, if the id would not be `13`, the node could not register at RouDi and `create` would return an error.

I added a test to ensure a node cannot register with a RouDi with a non-matching domain ID.

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp [hunk 11]
```diff
@@ -105,6 +106,115 @@ TEST(Node_test, CreatingNodeWithInvalidNameLeadsToError)
...
+
+    RouDiEnv roudi{DomainId{42}};
+
+    constexpr int32_t OVERWRITE_ENV_VARIABLE{1};
+    IOX_POSIX_CALL(setenv)
+    ("IOX_DOMAIN_ID", "42", OVERWRITE_ENV_VARIABLE)
+        .failureReturnValue(-1)
+        .evaluate()
+        .expect("Setting environment variable works!");
+    auto node_result = RouDiEnvNodeBuilder("foo").domain_id_from_env_or_default().create();
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:199]: see above

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp [hunk 12]
```diff
@@ -105,6 +106,115 @@ TEST(Node_test, CreatingNodeWithInvalidNameLeadsToError)
...
+
+TEST(Node_test, CreatingNodeWithDomainIdFromEnvOrDefaultWorksIfDomainIdIsNotSet)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "363dfb49-75fa-4486-b8b1-0f31c16bf37c");
+
+    RouDiEnv roudi{DEFAULT_DOMAIN_ID};
+
+    IOX_POSIX_CALL(unsetenv)
+    ("IOX_DOMAIN_ID").failureReturnValue(-1).evaluate().expect("Unsetting environment variable works!");
+    auto node_result = RouDiEnvNodeBuilder("foo").domain_id_from_env_or_default().create();
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:212]: see above

File: iceoryx_posh/source/runtime/shared_memory_user.cpp [hunk 13]
```diff
@@ -22,121 +22,128 @@
...
+{
+    constexpr access_rights SHM_SEGMENT_PERMISSIONS =
+        perms::owner_read | perms::owner_write | perms::group_read | perms::group_write;
+
+    auto shmResult = PosixSharedMemoryObjectBuilder()
+                         .name(concatenate(iceoryxResourcePrefix(domainId, resourceType), shmName))
+                         .memorySizeInBytes(shmSize)
+                         .accessMode(accessMode)
+                         .openMode(OpenMode::OPEN_EXISTING)
+                         .permissions(SHM_SEGMENT_PERMISSIONS)
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/runtime/shared_memory_user.cpp:121]: Can you elaborate on this? I did not change the parameters, they are the same like before.

File: iceoryx_hoofs/test/moduletests/test_memory_scope_guard.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_hoofs/test/moduletests/test_memory_scope_guard.cpp:159]: Yes, this would be fine.

File: iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp]: But with this construct, you could add `explicit` in front of the ctor.

File: iceoryx_posh/source/runtime/shared_memory_user.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_posh/source/runtime/shared_memory_user.cpp:121]: The permission relate to the `mode` parameter in `shm_open(const char *name, int oflag, mode_t mode);` that sets the permissions on creation when `O_CREAT` is set and the shared memory is created, otherwise it is ignored.

File: iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/experimental/include/iox/posh/experimental/node.hpp]: That's already the case and therefore it is `domain_id(DomainId{42})` instead of just `domain_id(42)`

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (elfenpiff) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:199]: We have generic newtype tests?

File: iceoryx_posh/source/runtime/shared_memory_user.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/source/runtime/shared_memory_user.cpp:121]: Excuse my ignorance but do you suggest to just remove `.permissions(...)`?

File: iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_posh/test/integrationtests/test_posh_experimental_node.cpp:199]: Yes, they are in `test_design_newtype.cpp`

**Review by elfenpiff — APPROVED** [review]

**Review by elfenpiff — APPROVED** [review]
